File: Fracture/Fracture/Fractal/CoordinateTransformer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Fracture.Fractal
{
    public class CoordinateTransformer
    {
        public double ExternalCenterX { get; set; }
        public double ExternalCenterY { get; set; }

        public double LocalCenterX { get; set; }
        public double LocalCenterY { get; set; }

        public double XRatio { get; set; }
        public double YRatio { get; set; }

        public Tuple<double, double> TransformToLocal(double x, double y)
        {
            return TransformToLocal(Tuple.Create(x, y));
        }

        public Tuple<double, double> TransformToLocal(Tuple<double, double> point)
        {
            var localX = (point.Item1 - ExternalCenterX) * XRatio + LocalCenterX;
            var localY = (point.Item2 - ExternalCenterY) * YRatio + LocalCenterY;
            return Tuple.Create(localX, localY);
        }

        public Tuple<double, double> TransformToExternal(double x, double y)
        {
            return TransformToExternal(Tuple.Create(x, y));
        }

        public Tuple<double, double> TransformToExternal(Tuple<double, double> point)
        {
            var externalX = (point.Item1 - LocalCenterX) / XRatio + ExternalCenterX;
            var externalY = (point.Item2 - LocalCenterY) / YRatio + ExternalCenterY;
            return Tuple.Create(externalX, externalY);
        }

        public static CoordinateTransformer CreatePixelTransformer(
            int pixWidth,
            int pixHeight,
            double logicalXOrigin,
            double logicalYOrigin,
            double logicalXWidth)
        {
            return new CoordinateTransformer
            {
                ExternalCenterX = (double)pixWidth / 2.0,
                ExternalCenterY = (double)pixHeight / 2.0,
                XRatio = logicalXWidth / pixWidth,
                YRatio = -(logicalXWidth / pixWidth),
                LocalCenterX = logicalXOrigin,
                LocalCenterY = logicalYOrigin
            };
        }
    }
}

File: Fracture/fracture-ui/src/util/CoordinateTransformer.js

export default class CoordinateTransformer {

  constructor(
      externalCenterX, externalCenterY,
      localCenterX, localCenterY,
      xRatio, yRatio)
  {
    this.externalCenterX = externalCenterX;
    this.externalCenterY = externalCenterY;
    this.localCenterX = localCenterX;
    this.localCenterY = localCenterY;
    this.xRatio = xRatio;
    this.yRatio = yRatio;
  }

  toLocal(x, y) {
    var localX = (x - this.externalCenterX) * this.xRatio + this.localCenterX;
    var localY = (y - this.externalCenterY) * this.yRatio + this.localCenterY;
    return [localX, localY];
  }

  toExternal(x, y) {
    var externalX = (x - this.localCenterX) / this.xRatio + this.externalCenterX;
    var externalY = (y - this.localCenterY) / this.yRatio + this.externalCenterY;
    return [externalX, externalY];
  }

  static createPixelTransformer(
    pixWidth,
    pixHeight,
    logicalXOrigin,
    logicalYOrigin,
    logicalWidth)
  {
    return new CoordinateTransformer(
        pixWidth / 2.0,
        pixHeight / 2.0,
        logicalXOrigin,
        logicalYOrigin,
        logicalWidth / pixWidth,
        -(logicalWidth / pixWidth),
    );
  }
};
